tiledb/api/src/query/subarray.rs

```diff
@@ -33,6 +39,121 @@ pub struct Subarray<'ctx> {
     raw: RawSubarray,
 }
 
+impl<'ctx> Subarray<'ctx> {
+    pub(crate) fn capi(&self) -> *mut ffi::tiledb_subarray_t {
+        *self.raw
+    }
+
+    pub(crate) fn new(context: &'ctx Context, raw: RawSubarray) -> Self {
+        Subarray { context, raw }
+    }
+
+    /// Return all dimension ranges set on the query.
+    pub fn ranges(
+        &self,
+        schema: &'ctx Schema,
```

Comment:
Can a subarray exist without a query?  Maybe instead of an `&'ctx Context` member subarray should have a `&'query Query` member.  THen you can fish the array out of the query and fish the schema out of the array.

Reply:
Huh! Good idea! The [`tiledb_subarray_alloc` requires an `tiledb_array_t*`](https://github.com/TileDB-Inc/TileDB/blob/dev/tiledb/sm/c_api/tiledb.cc#L1739-L1742) so shoving that into the struct instead seems like a good approach.

Reply:
This actually turned out more difficult than expected. It feels like it should work, but something about getting a schema from the array on a query is causing lifetime issues:

Code:
https://github.com/TileDB-Inc/tiledb-rs/commit/b697050067c4e92cc144f79dd601d41e0ef5b992

Compiler Error:
https://gist.github.com/davisp/10755108b147a004825eb76193db4a3d

I'll try fixing this later but might end up abandoning it since it might all be moot if we remove 'ctx lifetimes.

Reply:
[I get by with a little help from my friends.](https://github.com/TileDB-Inc/tiledb-rs/pull/72/commits/f4cbf19767080134d533506d18b58067622c41b6)

Thanks for the assist!